services/gundeck/src/Gundeck/API/Public.hs

```diff
@@ -16,185 +16,56 @@
-   in pure $ json doc
-
-addTokenH :: UserId ::: ConnId ::: JsonRequest Public.PushToken ::: JSON -> Gundeck Response
-addTokenH (uid ::: cid ::: req ::: _) = do
-  newtok <- fromJsonBody req
-  handleAddTokenResponse <$> Push.addToken uid cid newtok
-
-handleAddTokenResponse :: Push.AddTokenResponse -> Response
-handleAddTokenResponse = \case
-  Push.AddTokenSuccess newtok -> success newtok
-  Push.AddTokenNoBudget -> snsThreadBudgetReached
-  Push.AddTokenNotFound -> notFound
-  Push.AddTokenInvalid -> invalidToken
-  Push.AddTokenTooLong -> tokenTooLong
-  Push.AddTokenMetadataTooLong -> metadataTooLong
-
-success :: Public.PushToken -> Response
-success t =
-  let loc = Text.encodeUtf8 . Public.tokenText $ t ^. Public.token
-   in json t & setStatus status201 & addHeader hLocation loc
-
-invalidToken :: Response
-invalidToken =
-  json (mkError status400 "invalid-token" "Invalid push token")
-    & setStatus status404
-
-snsThreadBudgetReached :: Response
-snsThreadBudgetReached =
-  json (mkError status400 "sns-thread-budget-reached" "Too many concurrent calls to SNS; is SNS down?")
-    & setStatus status413
-
-tokenTooLong :: Response
-tokenTooLong =
-  json (mkError status400 "token-too-long" "Push token length must be < 8192 for GCM or 400 for APNS")
-    & setStatus status413
-
-metadataTooLong :: Response
-metadataTooLong =
-  json (mkError status400 "metadata-too-long" "Tried to add token to endpoint resulting in metadata length > 2048")
-    & setStatus status413
-
-notFound :: Response
-notFound = empty & setStatus status404
-
-deleteTokenH :: UserId ::: Public.Token ::: JSON -> Gundeck Response
-deleteTokenH (uid ::: tok ::: _) =
-  setStatus status204 empty <$ Push.deleteToken uid tok
-
-listTokensH :: UserId ::: JSON -> Gundeck Response
-listTokensH (uid ::: _) =
-  setStatus status200 . json @Public.PushTokenList <$> Push.listTokens uid
+servantSitemap :: ServerT GundeckAPI Gundeck
+servantSitemap = pushAPI :<|> notificationAPI
+  where
+    pushAPI =
+      Named @"register-push-token" addToken
+        :<|> Named @"delete-push-token" deleteToken
+        :<|> Named @"get-push-tokens" listTokens
+
+    notificationAPI =
+      Named @"get-notification-by-id" getById
+        :<|> Named @"get-last-notification" getLastNotification
+        :<|> Named @"get-notifications@v2" paginateUntilV2
+        :<|> Named @"get-notifications" paginate
+
+addToken :: UserId -> ConnId -> Public.PushToken -> Gundeck (Either Public.AddTokenError Public.AddTokenSuccess)
+addToken uid cid newTok =
+  Push.addToken uid cid newTok <&> \case
+    Push.AddTokenSuccess t -> Right (Public.AddTokenSuccess t)
+    Push.AddTokenNoBudget -> Left Public.AddTokenErrorNoBudget
+    Push.AddTokenNotFound -> Left Public.AddTokenErrorNotFound
+    Push.AddTokenInvalid -> Left Public.AddTokenErrorInvalid
+    Push.AddTokenTooLong -> Left Public.AddTokenErrorTooLong
+    Push.AddTokenMetadataTooLong -> Left Public.AddTokenErrorMetadataTooLong
```

Comment:
would it have made sense to use `Public.AddTokenSuccess` in `Push` rather than introducing two completely identical ADTs?